azhttpclient/middleware.go

```diff
@@ -27,31 +28,49 @@ func AzureMiddleware(authOpts *AuthOptions, credentials azcredentials.AzureCrede
 			return errorResponse(err)
 		}
 
+		if authOpts.rateLimitSession {
+			sessionProvider, err = newSessionProvider()
+			if err != nil {
+				return errorResponse(err)
+			}
+		}
+
 		if len(authOpts.scopes) == 0 {
 			err = errors.New("scopes not configured")
 			return errorResponse(err)
 		}
 
-		return applyAzureAuth(tokenProvider, authOpts.scopes, authOpts.endpoints, next)
+		return applyAzureAuth(tokenProvider, sessionProvider, authOpts.scopes, authOpts.endpoints, next)
 	})
 }
 
-func applyAzureAuth(tokenProvider aztokenprovider.AzureTokenProvider, scopes []string,
-	endpoints *azendpoint.EndpointAllowlist, next http.RoundTripper) http.RoundTripper {
+func applyAzureAuth(tokenProvider aztokenprovider.AzureTokenProvider, sessionProvider *userSessionProvider,
+	scopes []string, endpoints *azendpoint.EndpointAllowlist, next http.RoundTripper) http.RoundTripper {
 	return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+		reqContext := req.Context()
+
 		if endpoints != nil {
 			endpoint := azendpoint.Endpoint(*req.URL)
 			if !endpoints.IsAllowed(endpoint) {
 				return nil, fmt.Errorf("request to endpoint '%s' is not allowed by the datasource", endpoint.String())
 			}
 		}
 
-		token, err := tokenProvider.GetAccessToken(req.Context(), scopes)
+		token, err := tokenProvider.GetAccessToken(reqContext, scopes)
 		if err != nil {
 			return nil, fmt.Errorf("failed to retrieve Azure access token: %w", err)
 		}
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
 
+		if sessionProvider != nil {
+			sessionId, err := sessionProvider.GetSessionId(reqContext)
+			if err != nil {
+				return nil, fmt.Errorf("failed to obtain user session: %w", err)
```

Comment:
I think we should log the error here but not fail the request if there is an error generating the `sessionId` value.

Reply:
This error is only going to appear due to developer error, improper configuration of datasource http client.
e.g. when the rate limit session logic was enabled in code but user context wasn't configured.

It isn't supposed to transiently appear at runtime.